crates/sparrow-runtime/src/execute.rs

```diff
@@ -200,23 +188,58 @@ pub async fn execute_new(
             .into_report()
             .change_context(Error::internal_msg("decode primary_grouping_key_type"))?;
 
-    let mut key_hash_inverse = KeyHashInverse::from_data_type(primary_grouping_key_type.clone());
+    let primary_group_id = data_context
+        .get_or_create_group_id(&plan.primary_grouping, &primary_grouping_key_type)
+        .into_report()
+        .change_context(Error::internal_msg("get primary grouping ID"))?;
+
+    let mut key_hash_inverse = KeyHashInverse::from_data_type(&primary_grouping_key_type.clone());
     if let Some(compute_store) = &compute_store {
         if let Ok(restored) = KeyHashInverse::restore_from(compute_store.store_ref()) {
             key_hash_inverse = restored
         }
     }
 
-    let primary_group_id = data_context
-        .get_or_create_group_id(&plan.primary_grouping, &primary_grouping_key_type)
-        .into_report()
-        .change_context(Error::internal_msg("get primary grouping ID"))?;
-
     key_hash_inverse
-        .add_from_data_context(&data_context, primary_group_id, &object_stores)
+        .add_from_data_context(data_context, primary_group_id, object_stores)
         .await
         .change_context(Error::internal_msg("initialize key hash inverse"))?;
     let key_hash_inverse = Arc::new(ThreadSafeKeyHashInverse::new(key_hash_inverse));
+    Ok(key_hash_inverse)
+}
+
+/// Execute a given query based on the options.
+///
+/// Parameters
+/// ----------
+/// - key_hash_inverse: If set, specifies the key hash inverses to use. If None, the
+///   key hashes will be created.
+pub async fn execute_new(
+    plan: ComputePlan,
+    destination: Destination,
+    mut data_context: DataContext,
+    options: ExecutionOptions,
+    key_hash_inverses: Option<Arc<ThreadSafeKeyHashInverse>>,
```

Comment:
make this singular?

Reply:
try to be consistent with other place (we use them interchangeably)